### ISSUE TITLE
[Process] update the vm running ota.koreader.rocks

### ISSUE THREAD

**Issue by Hzj-jie** [koreader/koreader#10615]
Now it's a 1st gen vm on azure, the vm type has been deprecated and would be removed around 2024.
Meanwhile the Ubuntu is 16, and should be upgraded.

**Comment by Hzj-jie** [comment]
I have some time recently to work on it.
But before starting, do we still have a list of the binaries in the machine?

**Comment by Frenzie** [comment]
How do you mean precisely? Like the list of apt installed packages and the Docker images?

Wrt the OS we shouldn't need much of anything other than Bash and Docker; all the important things are done in Docker.

It might be quicker to talk on Gitter or something btw, if you're on there?

**Comment by Hzj-jie** [comment]
Oh, that's great. Previously I thought we copied the binaries.
When would be the good time to do the upgrade? I haven't done it before, I am not very sure if I can preserve the original ip.
I will do some homework first.

**Comment by Frenzie** [comment]
There are a couple of things on there like `ncdu` that can be useful but I don't think a specific effort is needed there. We can just install it if or when we need it. What's important is ops and the user configs.

Pinging @houqp for the Cloudflare stuff.

**Comment by Hzj-jie** [comment]
I know progress sync and ota are on the machine, while progress sync has persistent date. Anything else?

**Comment by Frenzie** [comment]
The configs and scripts, the signing key, I meant the ops thing quite literally (perhaps including the actual archives for ota, though that'd regenerate in time). Just everything that's in there.

What's in the cronfile is potentially also important, though it's probably just to run the cleanup script once or twice a day.

**Comment by Hzj-jie** [comment]
OK, let me first upgrade the machine type, currently it's running an A1_Basic which will be deprecated August 2024.
https://learn.microsoft.com/en-us/azure/virtual-machines/av1-series-retirement

VM | Size | Type | vCPUs | RAM (GiB) | Data disks | Max IOPS | Temp storage (GiB) | Premium disk | Cost/month
A1_Basic | General purpose | 1 | 1.75 | 2 | 2x500 | 40 | Not supported | $22.63
I.e. 1 vCPU 1.75G ram, 1000 IOPS

Likely the B1ms VM type is the best approach.
B1ms | General purpose | 1 | 2 | 2 | 640 | 4 | Supported | $18.10
I.e. 1 vCPU 2G ram and 640 IOPS

The only thing matters is the temporary storage, 40GB vs 4GB, not sure if it's sufficient.

Or if 1G ram is enough, we can try B1s, and save more budget for the bandwidth - azure charges bandwidth separately which was the reason it ran out my budget from the subscription.
B1s | General purpose | 1 | 1 | 2 | 320 | 4 | Supported | $9.05
I.e. 1G ram and 320 IOPS vs 2G ram and 640 IOPS.

It seems like I lost access to the VM, previously it used a user name & password. But now it accepts only the ssh key-pair. But azure is very bad at the keys, replacing the key would break you guys I believe.
If possible, would you mind to add my public key so that I can log into it and take a look.

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDfnbW0jcCTFJknPks6Lir9ZZfiX8By62414r0bvN4ciIQWleU147Ma4ZBrR5E7GV8IyX4zbLmldI00uKbFb9q2IpHN7ebNmKfIOnDnTFOuLMPjsAUHCl13yIr0yLlEWILu1Tni7w3oeNXGy7WK2oDJ5DgwASvoCrQ5GKg3SNnmBJBk/1EMWVQHZ+SralMx5Udz80ij1YutWV0S8kJH3YgHXE1G4SVmTq9oC7riMI5l1QWJgaKynrY2D171VRhIbafqLkR7SmcN1Vw23fnAEbIga94SBcXyl9tVG2r5rYUSGyHkvO+rjc+XHf701AudeG/+LQB2Uf5t90y8e1oV5IREeVz2BIYomQBNQjTyS+EoOB+ai/ponXwaeoVUxTdYRWQTzVtZ0ewPoLTheCK8qEDwm04al4xjWgfCAXzd+5ZKpS2NcgLi+RDL3bN5uExMPevOtkVhhr1BrdJFBz7LeB1X7cx34I1GxhuLfGMoh0oVYhmv8dC2RIoNCngRzFVTKq0= hzj-jie@hzj-jie-x1

**Comment by Hzj-jie** [comment]
By the way, changing the machine type (size in azure concept) would preserve the data but restart the machine. Should we set up the services into systemd first?

**Comment by Frenzie** [comment]
I'm thinking by Docker container name, something like this?
```
[Unit]
Description=Docker nginx
Requires=docker.service
After=docker.service

[Service]
Restart=always
ExecStart=/usr/bin/docker start -a nginx
ExecStop=/usr/bin/docker stop -t 2 nginx

[Install]
WantedBy=multi-user.target
```

**Comment by Frenzie** [comment]
> The only thing matters is the temporary storage, 40GB vs 4GB, not sure if it's sufficient.

What does temporary storage mean exactly? Over in /mnt? That's not really used atm; actually I'm not sure I even knew that existed. I figure we can easily make do with 4 GB for some temp extraction/file manipulation provided the actually important permanent stuff is at least some 30 GB as it is now.

**Comment by Hzj-jie** [comment]
I don't know, but I can take a look once I can login to the machine.

**Comment by poire-z** [comment]
I've just added your key into your own authorized_keys (hoping you remember the spelling of your username :)
(Been years since I last logged into this server :)

**Comment by Hzj-jie** [comment]
I cleaned up some my bad user names. But if I unfortunately broke your authentication, let me know.

**Comment by Hzj-jie** [comment]
So the machine has two disks. The root is sda1, persistent AFAICT, and it's ~30G; the sdb is mounted at /mnt, temporary, and it's ~40G. So I think the "Temporary storage" means it, but we never use it.

```
/dev/sdb1       41151808    49036  38989340   1% /mnt

Hzj_jie@hzj-jie-ubuntu:/$ ll /mnt/
total 28
drwxr-xr-x  3 root root  4096 Jun 26 02:24 ./
drwxr-xr-x 25 root root  4096 Jun 26 02:23 ../
-r--r--r--  1 root root   639 Jun 26 02:24 DATALOSS_WARNING_README.txt
drwx------  2 root root 16384 Jun 26 02:23 lost+found/
```

The sda1 in contrast is 98% full,
```
/dev/sda1       29711408 29068328    626696  98% /

Hzj_jie@hzj-jie-ubuntu:/$ du -h -d 1 2>/dev/null
44K     ./tmp
680K    ./run
4.0K    ./root
17M     ./bin
14M     ./sbin
35M     ./opt
0       ./dev
159M    ./boot
8.0K    ./snap
4.0K    ./lib64
952M    ./lib
7.7M    ./etc
0       ./proc
249M    ./home
17G     ./ops
4.0K    ./docker
16K     ./lost+found
1.3G    ./usr
0       ./sys
24K     ./mnt
4.0K    ./srv
4.0K    ./media
438M    ./var
20G     .
```

Pretty much everything is in /ops, or more specifically, /ops/prod/build/download/stable. Are we serving only the latest build from here via ota? If so we can delete the old stable versions then.

For the server update, it's lucky that the do-release-upgrade exists on the vm and we may use it.

**Comment by Frenzie** [comment]
>  If so we can delete the old stable versions then.

Not as a matter of course, people depend on that and we purposefully keep a limited number of old ones around. But in order to free up 10+ GB for a release upgrade, go right ahead. :-)

**Comment by Frenzie** [comment]
Same story for the older nightlies.

**Comment by Frenzie** [comment]
PS I'm sure you're aware, but don't forget to run the upgrade in tmux in case of connectivity issues. :-)

**Comment by Hzj-jie** [comment]
I will try :)

**Comment by Hzj-jie** [comment]
Added and enabled services:

```
Hzj_jie@hzj-jie-ubuntu:/ops/prod/build/download$ sudo systemctl enable kosync.service
Created symlink from /etc/systemd/system/multi-user.target.wants/kosync.service to /etc/systemd/system/kosync.service.
Hzj_jie@hzj-jie-ubuntu:/ops/prod/build/download$ sudo systemctl enable kosync.service
Hzj_jie@hzj-jie-ubuntu:/ops/prod/build/download$ sudo systemctl enable nginx
Created symlink from /etc/systemd/system/multi-user.target.wants/nginx.service to /etc/systemd/system/nginx.service.
Hzj_jie@hzj-jie-ubuntu:/ops/prod/build/download$ sudo systemctl enable nightswatcher
Created symlink from /etc/systemd/system/multi-user.target.wants/nightswatcher.service to /etc/systemd/system/nightswatcher.service.
```

Anyone want to confirm if I have done it right?

**Comment by Frenzie** [comment]
Luckily you don't seem to have interfered with my release proceedings. :sweat_smile: 

I'll check in a few minutes.

**Comment by Hzj-jie** [comment]
Oh, I will definitely let you guys know first before proceeding updates or restarting.

**Comment by Frenzie** [comment]
I adjusted the nightswatcher startup script which had `--rm` in it (so stop would remove it immediately) but I think it's good to go. I also temporarily removed all the nightly/stable files that are merely nice to have and not crucial, so I figure there should be enough space for a release upgrade.

```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            821M     0  821M   0% /dev
tmpfs           169M   18M  151M  11% /run
/dev/sda1        29G   10G   19G  36% /
tmpfs           842M  368K  842M   1% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           842M     0  842M   0% /sys/fs/cgroup
none             64K     0   64K   0% /etc/network/interfaces.dynamic.d
/dev/sdb1        40G   48M   38G   1% /mnt
tmpfs           169M     0  169M   0% /run/user/1004
tmpfs           169M     0  169M   0% /run/user/1002
```

**Comment by Frenzie** [comment]
Incidentally I just noticed a Prometheus and Grafana that haven't run in years. Therefore no need to add it to SystemD at this time, but just mentioning it.

![image](https://github.com/koreader/koreader/assets/202757/25881720-56b0-41dd-8761-70558044f981)

**Comment by Hzj-jie** [comment]
I created https://github.com/koreader/koreader-misc/pull/45 to add the definition files into koreader-misc.

Meanwhile, what's the good time for me to try a restart and ensure the services would be up and running after the restart?

**Comment by Frenzie** [comment]
It doesn't really matter if the nightlies are missing for a day, but the ideal time would be sometime after 7 UTC so the artifacts for the day are there already. Unless you mean wrt the current Android tc update (e.g., #10679) in which case it may be better to wait until that's sorted out.

**Comment by Hzj-jie** [comment]
Oh, maybe in another way, would you please share the commands you are using to start the services? I do see five dockers now. My concern is that if the services are not working as expected (I do not think so though), I can manually start the dockers to avoid breaking the services.

**Comment by Hzj-jie** [comment]
By the way, memory-wise, 1G would be a little bit restricted if I read /proc/meminfo correctly.

```
root@hzj-jie-ubuntu:~# cat /proc/meminfo
MemTotal:        1724096 kB
MemFree:          157712 kB
MemAvailable:     991028 kB
Buffers:           49364 kB
Cached:           964028 kB
SwapCached:            0 kB
Active:           952812 kB
Inactive:         424736 kB
```

**Comment by Frenzie** [comment]
> Oh, maybe in another way, would you please share the commands you are using to start the services?

They're scripts in ops (under build, nginx and sync).

**Comment by Hzj-jie** [comment]
OK, that's great. I will now restart the machine to take a shot.

**Comment by Hzj-jie** [comment]
Sorry, I just noticed your comment on https://github.com/koreader/koreader-misc/pull/46. I disabled grafana.service and prom.service.

**Comment by Hzj-jie** [comment]
The server has just been restarted. I am monitoring the restart process.

**Comment by Hzj-jie** [comment]
Three services nightswatcher nginx and kosync are all up and running.
https://ota.koreader.rocks/ is working; ota and progress sync on my koreader are also both working.

**Comment by Hzj-jie** [comment]
I will go ahead to update the machine spec tomorrow if you have no concerns.

**Comment by Frenzie** [comment]
Sounds good! Just to double check, that'll keep the IP?

**Comment by Hzj-jie** [comment]
I think so, let me confirm.

**Comment by Frenzie** [comment]
Btw, I thought you wanted to run a do-release-upgrade first or did I misunderstand?

**Comment by Hzj-jie** [comment]
Previously I planned to resize the machine first. Is there a reason to upgrade the ubuntu first?

**Comment by Hzj-jie** [comment]
The "official" use of the temp drive is virtual memory I think: https://learn.microsoft.com/en-us/azure/virtual-machines/azure-vms-no-temp-disk#how-do-i-migrate-from-a-vm-size-with-local-temp-disk-to-a-vm-size-with-no-local-temp-disk---.

**Comment by Frenzie** [comment]
> Is there a reason to upgrade the ubuntu first?

I thought you'd mentioned something about upgrading it first because the new config didn't support 16.04 anymore or something. For me it makes no difference.

**Comment by Hzj-jie** [comment]
Sorry, I was a little bit busy these days, I will start the resize process.

**Comment by Hzj-jie** [comment]
Done, and dockers are up and running. Go ahead to update the ubuntu itself.

**Comment by Hzj-jie** [comment]
On bionic / 18.04 now.

I manually updated /etc/sysctl.conf to enable `net.ipv4.ip_forward=1`. I do believe it was previously manually enabled for docker instances. /etc/default/ufw is also updated to `DEFAULT_FORWARD_POLICY="ACCEPT"`.
/etc/ssh/sshd_config was heavily modified but I do not see anything unexpected, so I use the apt version.

```
root@hzj-jie-ubuntu:~# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.6 LTS
Release:        18.04
Codename:       bionic
root@hzj-jie-ubuntu:~# docker ps
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS              PORTS                                      NAMES
8a21bda15a62        koreader/nightswatcher:0.8.1   "/usr/local/bin/gu..."   8 days ago          Up 58 seconds       0.0.0.0:9742->9742/tcp                     nightswatcher
0ab375aa4c04        frenzie/nginx-extras:1.21      "nginx -g 'daemon ..."   2 weeks ago         Up 57 seconds       0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   nginx
92965ffac5e7        d1fcf99cb270                   "/sbin/my_init"          3 weeks ago         Up 58 seconds       0.0.0.0:7200->7200/tcp                     kosync
```

**Comment by Hzj-jie** [comment]
On focal / 20.04 now. Same changes above.

```
root@hzj-jie-ubuntu:~# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal
root@hzj-jie-ubuntu:~# docker ps
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS              PORTS                                      NAMES
8a21bda15a62        koreader/nightswatcher:0.8.1   "/usr/local/bin/gu..."   8 days ago          Up 39 seconds       0.0.0.0:9742->9742/tcp                     nightswatcher
0ab375aa4c04        frenzie/nginx-extras:1.21      "nginx -g 'daemon ..."   2 weeks ago         Up 39 seconds       0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   nginx
92965ffac5e7        d1fcf99cb270                   "/sbin/my_init"          3 weeks ago         Up 39 seconds       0.0.0.0:7200->7200/tcp                     kosync
```

We can move forward to 22.04 if necessary.

**Comment by Hzj-jie** [comment]
Enabled /mnt/swapfile as swap.

**Comment by Hzj-jie** [comment]
OK, go ahead to update to 22.04.

**Comment by Hzj-jie** [comment]
The latest docker disallows the use of aufs, should we update to overlay2?

**Comment by Frenzie** [comment]
I'm surprised it doesn't migrate by itself if that's the case. But sure, I doubt that makes a difference? Worst case scenario we can just redownload the images from Docker Hub so they'll use overlay2.

**Comment by Hzj-jie** [comment]
Oops bandwidth usage exceeded the limit again.

**Comment by Hzj-jie** [comment]
Our network usage is several hundreds MB per hour, but sometimes 15GB per hour. I suspect some crawlers crawling the site. Maybe a robot.txt would help if we don't have one.

**Comment by Frenzie** [comment]
I don't see how that's remotely possible. Cloudflare sits in between and it should only have cache misses at a rate of a few hundred MB per day (i.e., new nightlies). It looks like this means I also can't get in to analyze the logs.

Pinging @houqp

**Comment by Hzj-jie** [comment]
Unfortunately the monitoring shows only the data from the last 30 days, I cannot tell when it started happening. I can inspect the nginx log after restarting the machine on 26th.

**Comment by Hzj-jie** [comment]
The vm has been restarted.

**Comment by Frenzie** [comment]
https://github.com/koreader/koreader/issues/10747#issuecomment-1650607082 reminded me that the sync server could be to blame (if spambots are on it or something)

**Comment by Hzj-jie** [comment]
Not sure if I can rely on the format of ngnix log, but I got the following numbers:

```
/var/log/nginx # for i in $(ls -1) ; do echo $i ; cut -d ' ' -f 10 $i | awk '{ a += $1 } END { print a }' ; done
access.log
11364276
build.access.log
4143660874586
default.access.log
201626617
error.log
140.162
kosync.access.log
386626134
ota.access.log
285717342362
```

Unfortunately, the logs start from different times. access, build.access, default.access and kosync.access start from ~Jun 2017. ota.access starts from July 09, 2023. But build.access and ota.access definitely contribute most of the traffic.

Then I ran the following command to get the line index starting from 09/Jul/2023.
```
/var/log/nginx # for i in $(ls -1) ; do echo $i ; grep -n 09/Jul/2023 $i | head -n 1; done
access.log
build.access.log
1308988:172.70.39.9 - - [09/Jul/2023:00:20:47 +0000] "GET / HTTP/1.1" 404 146 "http://build.koreader.rocks/" "Mozilla/5.0 (compatible; DuckDuckGo-Favicons-Bot/1.0; +http://duckduckgo.com)"
default.access.log
700770:213.109.202.66 - - [09/Jul/2023:00:11:15 +0000] "GET / HTTP/1.1" 404 548 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36"
error.log
kosync.access.log
3488034:162.158.146.239 - - [09/Jul/2023:00:03:46 +0000] "GET /syncs/progress/65a173ddc8ddba48db602b70cf3749e8 HTTP/1.1" 200 2 "-" "lua-Spore v0.3.3"
ota.access.log
1:172.70.46.203 - - [09/Jul/2023:08:51:55 +0000] "GET /koreader-kobo-latest-stable.zsync HTTP/1.1" 200 587848 "-" "KOReader/2023.06-20 (https://koreader.rocks/) LuaSocket/3.0.0"
```

Then
```
/var/log/nginx # tail -n +1308988 build.access.log | cut -d ' ' -f 10 | awk '{ a += $1 } END { print a }'
36031612668
```

So ota is the one contributing the most traffic. I think its traffic in general cannot be cached unless the proxy is smart enough to automatically understand the concept of 206.

**Comment by Frenzie** [comment]
> ota.access starts from July 09, 2023. 

Yes, I deleted it then because it was somewhat large and space was needed for upgrades.

>  I think its traffic in general cannot be cached unless the proxy is smart enough to automatically understand the concept of 206.

Afaik Cloudflare doesn't cache partial content requests per se but it downloads (and subsequently caches) the entire file to serve the partial content, if it doesn't fully have it yet.

As you said, clearly that's not the case (that's `goaccess`, available in the repos). But notably APK and AppImage *also* use many gigabytes. So clearly Cloudflare has cache misses… constantly? That's 19.78 GB for koreader-android-arm-v2023.06.1.apk.

![image](https://github.com/koreader/koreader/assets/202757/2d56dcd5-8c53-48a2-9caf-44a95081ecd3)

**Comment by NiLuJe** [comment]
FWIW, switching to .tar.gz should be trivial (I never really understood the reasoning in the first place, but it was also related to ISP caches, but possibly only in China or something).

**Comment by NiLuJe** [comment]
And, also FWIW, I was at ~160GB of traffic for last June on my object storage containing the mirrors ;).

**Comment by NiLuJe** [comment]
> But I assume you don't have CF in front.

Yup, raw numbers there.

**Comment by Hzj-jie** [comment]
Interesting, caching results base on MIME or extension are both inaccurate. There is a cache-control and an expires http response headers.

**Comment by Frenzie** [comment]
Not to mention last-modified, but of course a CDN isn't really supposed to act like a normal browser. You generally want them to sync a lot sooner than you might want browser cache to expire.

**Comment by houqp** [comment]
sorry for the late response, i am looking into cloudflare now, hopefully i can transfer the ownership to the community as well.

**Comment by houqp** [comment]
I have turned on caching for `.targz` requests. Our cache hit rate is only about 20-30% based on cloudflare analytics FWIW.

I have also sent cloudflare invites to @Frenzie and @Hzj-jie so they can manage the domain going forward.

@NiLuJe could you send me your latest email so I can invite you as well?

**Comment by NiLuJe** [comment]
It's the email my GH commits are signed with, FWIW ;).

**Comment by Frenzie** [comment]
Is there a sane way to deal with this techbro circlejerk?

![image](https://github.com/koreader/koreader/assets/202757/58a893f9-24f8-40fd-b3b9-8d3e64905c81)

**Comment by NiLuJe** [comment]
> Is there a sane way to deal with this techbro circlejerk?

I let 1Password handle it, although the auto-detection didn't work, so it requires adding the hash manually.
(I'm assuming other password managers have started handling 2FA, which is hilarious, because that's not really 2FA anymore, if the exact same thing handles both the password and the OTP, but whatever. I keep an actual authenticator app for things that actually matter).

**Comment by Hzj-jie** [comment]
Yes, I got it.
But the vm is running out of quota again.

**Comment by Frenzie** [comment]
That would suggest the cache still wasn't being hit?

Hopefully I'll be able to check soon, but for the moment my conclusion is that Cloudflare is useless trash.

Note to self:
https://support.enpass.io/app/item/generating_one_time_code_in_enpass.htm

**Comment by Frenzie** [comment]
PS My conclusion on Azure is the same tbh.

**Comment by NiLuJe** [comment]
Speaking of CF, I noticed when working on #10839 that we're actually ending up with a CF timeout error code (522). Is that... expected? (I can't think of any reason where KOSync being behind CF would make any sense, but, this is admittedly out of my wheelhouse).

**Comment by Frenzie** [comment]
Probably? I think even when it's not caching it still routes it to edge nodes in some optimized manner which normally causes better speeds than doing all kinds of organic hopping.

**Comment by Frenzie** [comment]
So apparently we can't see right now what's being missed from CF, but the cache ratio does seem to have improved a bit before it went dark. I'll probably just add all the relevant extensions as an `OR` to the targz rule.
![image](https://github.com/koreader/koreader/assets/202757/29006b93-5ed7-48c3-ba80-4d587df9850b)

**Comment by Hzj-jie** [comment]
I just restarted the VM. Yes, the charge ratio of traffic is very high.

**Comment by Frenzie** [comment]
The logs are giving me the distinct feeling that .tar.gz and .apk aren't being cached either, despite supposedly being cached by default.

Unfortunately CF won't let me change the cache rules… I've sent @houqp an email about it.

![image](https://github.com/koreader/koreader/assets/202757/ea424ae5-0060-49d6-b2f2-6b3ba25f630a)

**Comment by NiLuJe** [comment]
AFAIK, that's probably because it's outside of the ToS for CF, they explicitly mention their default stock thingy is only there to cache web content (e.g., roughly html/css).

**Comment by NiLuJe** [comment]
As a quick fix, we could probably stuff the dictionaries elsewhere (e.g., my own stuff is on https://www.ovhcloud.com/fr/public-cloud/prices/#439 and I'm pretty sure you'd be hard-pressed to find cheaper BW anywhere else).

**Comment by Frenzie** [comment]
But as I said above when linking https://developers.cloudflare.com/cache/concepts/default-cache-behavior/ .apk and .tar.gz are explicitly listed as cached by default.

There's no need for a quick fix at this time; it's over 60% cached vs barely more than 30% before, and caching .targz came too late to make enough of an impact last month. It just should be higher still.

**Comment by Frenzie** [comment]
> To cache additional content [i.e., things other than APK, TAR.GZ and ZIP!], refer to [Page Rules](https://developers.cloudflare.com/cache/how-to/edge-browser-cache-ttl/create-page-rules/) to create a rule to cache everything.

**Comment by Frenzie** [comment]
> AFAIK, that's probably because it's outside of the ToS for CF, they explicitly mention their default stock thingy is only there to cache web content (e.g., roughly html/css).

Oh, I just realized you may have misinterpreted what I wrote. I meant CF was saying I don't have permission from the account owner, not that it said it won't do it.

**Comment by Frenzie** [comment]
> As a quick fix, we could probably stuff the dictionaries elsewhere (e.g., my own stuff is on https://www.ovhcloud.com/fr/public-cloud/prices/#439 and I'm pretty sure you'd be hard-pressed to find cheaper BW anywhere else).

Their hostile cookie dialog = saionara. But fwiw I pay $15/year for this (that bandwidth is per month) at <https://www.desivps.com/>, though I guess without an opportune sale/grandfathered account you can't get that for less than $30/year or some such. What you linked may potentially be cheaper under some 50-100 GB but you'd still need some way to interact with the dumb API anyway. So imho it sounds rather unattractive but ymmv. ^_^ Tbh I'd be more inclined to get a few of something like this and use GlusterFS or something, or just your traditional rsync cronjob.
![image](https://github.com/koreader/koreader/assets/202757/cf97e105-154a-41df-8bb7-c650f5b5cf1a)

And I pay $12/year for this over at Hostsailor; although I've had it for a decade I intend to retire that one. It's somehow still a lot faster than Azure though.
![image](https://github.com/koreader/koreader/assets/202757/14e36bc4-dff6-43a4-886b-8bb52af71ff7)

And yet…

Azure:
```
model name	: Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
```
El cheapo Hostsailor mesozoic server:
```
model name	: Intel(R) Xeon(R) CPU E31240 @ 3.30GHz
```

**Comment by Frenzie** [comment]
Caching rules updated to the following:

```
(ends_with(http.request.uri, ".apk")) or (ends_with(http.request.uri, ".AppImage")) or (ends_with(http.request.uri, ".targz")) or (ends_with(http.request.uri, ".tar.gz")) or (ends_with(http.request.uri, ".zip")) or (ends_with(http.request.uri, ".zsync"))
```

I guess we'll see tomorrow what impact it's had, if any.

**Comment by Hzj-jie** [comment]
There were spikes at least once a day before ~28th 9am UTC-7, i.e. 14h ago, but none since then. Though it doesn't match the time of the last comment, i.e. ~23h ago. Not sue if the rules would take effect immediately.

![image](https://github.com/koreader/koreader/assets/1188462/0fad6c5f-a0dd-481c-b1d8-e290402b514d)

**Comment by Frenzie** [comment]
In cached vs uncached bandwidth it looks positive compared to yesterday, but it's too early to spot any trends.

![image](https://github.com/koreader/koreader/assets/202757/ca89c77c-24a0-4766-9ede-da7f173553d0)

![image](https://github.com/koreader/koreader/assets/202757/0b9e2b09-f37a-4559-8de1-24fc3d5e92a1)

![image](https://github.com/koreader/koreader/assets/202757/c4d15776-c58f-4332-8750-a5b9ddb0e323)

**Comment by Hzj-jie** [comment]
OK, bad news -_-

Would it be possible to serve the updates directly from github?

![Screenshot_20230831-111621](https://github.com/koreader/koreader/assets/1188462/3cf8403a-d424-48c8-bdd4-bc4d2768c4f0)

**Comment by Frenzie** [comment]
What concretely is the bandwidth limit?

Those numbers seem to have little in common with these CF numbers for the past 7 (!) days:
![image](https://github.com/koreader/koreader/assets/202757/3aee3506-22b5-4f2c-8f56-e05cb461e07f)

The CF numbers don't count grabbing the nightly builds from GitLab, but that's no more than a few hundred MB.

**Comment by Hzj-jie** [comment]
Emmm, no idea. But the red lines above are outgoing traffic, so grabbing nightly shouldn't be a problem.

**Comment by Frenzie** [comment]
I thought 3 GB seemed like a bit more than expected, but not by a shocking amount. In any case your graph claims 70 GB in 3 days and CF claims only about 20 GB (uncached) in 7 days in the same period. I've installed vnstat to see what the machine itself says.

**Comment by Hzj-jie** [comment]
Running out of quota again, should we reduce the vm spec as a short term fix?

**Comment by Frenzie** [comment]
Already? :-/

I think we can safely rule out CF cache issues then seeing how only about 100 GB was uncached.
![image](https://github.com/koreader/koreader/assets/202757/d84678bf-b139-4878-9c0e-8b6f7f4b057f)

Or perhaps 100 GB is the (extremely low) value? I can't really tell what's applicable where.
<https://azure.microsoft.com/en-us/pricing/details/bandwidth/>
![image](https://github.com/koreader/koreader/assets/202757/6c927bdf-f352-42af-8b8a-818e97da418c)

> should we reduce the vm spec as a short term fix?

I'm not sure if I fully understand what you're saying with the spec. You mean the equivalent of adding a rate limit for the network?

Maybe the problem is something like this:
https://www.cyberciti.biz/faq/howto-limiting-ssh-connections-with-ufw-on-ubuntu-debian/

**Comment by Hzj-jie** [comment]
Oh, I meant to change the model of the vm to a lower level to shift some budget from the vm to network.

**Comment by Frenzie** [comment]
I think 1 GB should probably work but I'm not 100% sure.

**Comment by Hzj-jie** [comment]
Maybe we fall into this category.

![Screenshot_20230918-135531](https://github.com/koreader/koreader/assets/1188462/e5980661-aab2-43cd-a9a4-f3c76a69b16b)

Where is the CF server we are using? The azure machine is running in west us.

**Comment by Hzj-jie** [comment]
I resized the machine to 1 vCPU + 1G ram, and it seems working. Likely it can buy us 3-5 more days per month.

So the public address is public accessible, I am not sure if anyone crawl the site directly through the public ip.

**Comment by Hzj-jie** [comment]
I do see requests like

```
172.71.17.150 - - [15/Sep/2023:18:59:20 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09
-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64)
AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36 OPR/101.0.0.0"
```
and
```
162.158.175.177 - - [15/Sep/2023:20:53:55 +0000] "GET /download/nightly/v2023.03-45-ga8ab5e84_2023-04-18/ HTTP/1.1" 404 146 "-" "Mozilla/5.0 AppleWebKit/537.
36 (KHTML, like Gecko; compatible; GPTBot/1.0; +https://openai.com/gptbot)"
```

Both should be disallowed.

**Comment by Frenzie** [comment]
I can see why you'd say that about some bot (though it should either get it from the CF cache or be the first hence putting it in the cache) but the first is simply Opera so I'm not sure what you're getting at?

For reference, mine looks like this:
```
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 OPR/102.0.0.0
```

**Comment by Hzj-jie** [comment]
Oh, I got them from the nginx log.
I think the requests should be cached and only accessed by cf like

```
108.162.246.82 - - [04/Jun/2017:03:23:57 +0000] "HEAD / HTTP/1.1" 301 0 "-" "Mozilla/5.0 (compatible; CloudFlare-AlwaysOnline/1.0; +http://www.cloudflare.com/always-online)"
```

rather than the regular browsers.

**Comment by Frenzie** [comment]
A request like that will happen all the time for every caching edge node. There's no such thing as regular browsers just accessing, CF is always in between.

**Comment by Hzj-jie** [comment]
Oh, do you mean requests like

```
172.71.17.150 - - [15/Sep/2023:18:59:20 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09
-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64)
AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36 OPR/101.0.0.0"
```

are indeed passing through cf? I made a mistake here, the hosts are not resolved to the public ip address of the vm.

But as long as it returned 200, the request was served through the azure network and got charged. For example, the following apk was accessed 8 times, and some requests were hours apart, and cf should cache them.

```
172.70.91.197 - - [13/Sep/2023:10:05:47 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Linux; Android 11; RMX1901 Build/RKQ1.201217.002) AppleWebKit/537.36 (KHTML, like Gecko)  Chrome/116.0.0.0 Mobile Safari/537.36"
172.71.242.147 - - [13/Sep/2023:10:05:50 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Linux; Android 11; RMX1901 Build/RKQ1.201217.002) AppleWebKit/537.36 (KHTML, like Gecko)  Chrome/116.0.0.0 Mobile Safari/537.36"
172.70.86.27 - - [13/Sep/2023:10:06:01 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Linux; Android 11; RMX1901 Build/RKQ1.201217.002) AppleWebKit/537.36 (KHTML, like Gecko)  Chrome/116.0.0.0 Mobile Safari/537.36"
172.69.59.188 - - [14/Sep/2023:00:51:25 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Android 13; Mobile; rv:109.0) Gecko/117.0 Firefox/117.0"
172.71.186.230 - - [14/Sep/2023:01:40:37 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "http://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk" "Mozilla/5.0 (Android 13; Mobile; rv:68.0) Gecko/68.0 Firefox/68.0"
172.71.17.151 - - [14/Sep/2023:03:47:53 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"
172.68.10.185 - - [14/Sep/2023:03:49:10 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Mobile Safari/537.36"
172.71.17.150 - - [15/Sep/2023:18:59:20 +0000] "GET /download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/koreader-android-arm64-v2023.08-23-g3f677a7fd_2023-09-13.apk HTTP/1.1" 200 33851392 "https://build.koreader.rocks/download/nightly/v2023.08-23-g3f677a7fd_2023-09-13/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36 OPR/101.0.0.0"
```

Or the tar.gz files in the dict/
```
172.69.33.129 - - [26/Sep/2023:02:31:50 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.08 (https://koreader.rocks/) LuaSoc
ket/3.0.0"
162.158.91.27 - - [26/Sep/2023:02:31:51 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.08 (https://koreader.rocks/) LuaSoc
ket/3.0.0"
172.71.130.85 - - [26/Sep/2023:03:00:08 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.08-8028 (https://koreader.rocks/) L
uaSocket/3.0.0"
172.71.122.161 - - [26/Sep/2023:03:00:10 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.08-8028 (https://koreader.rocks/)
LuaSocket/3.0.0"
162.158.158.199 - - [26/Sep/2023:03:08:32 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.05.1 (https://koreader.rocks/) Lu
aSocket/3.0.0"
162.158.155.194 - - [26/Sep/2023:03:08:35 +0000] "GET /download/dict/gcide.tar.gz HTTP/1.1" 200 14958781 "-" "KOReader/2023.05.1 (https://koreader.rocks/) Lu
aSocket/3.0.0"
```

I haven't seen cache related headers in the nginx configuration,

```
        location /download {
                alias /data/release_download;
                autoindex on;
                fancyindex on;
                fancyindex_exact_size off;
                fancyindex_name_length 95;
                fancyindex_natural_sort on;
                fancyindex_header /fancyindex/header.html;
                fancyindex_ignore fancyindex;
        }
```

Are you sure cf would cache the files in the case?

**Comment by Frenzie** [comment]
> are indeed passing through cf?

Yes. It's impossible for it to be any other way.

> For example, the following apk was accessed 8 times, and some requests were hours apart, and cf should cache them.

"Hours apart" combined with different edge nodes means there's no particular reason it would be cached. They might drop it sooner than whatever you set or might expect if it hasn't been accessed in a while. Something like CF helps the most when there are a ton of requests coming in a short span using the same edge node. But that's why we have a 50-60% cache hit rate rather than over 90% (on average, this morning at 7 it was actually 96.54%).

> Are you sure cf would cache the files in the case?

Yes, you can easily verify for yourself with curl -I or -v.

Most important, you can see CF is aware of `Last-Modified: Wed, 09 Jan 2019 15:35:31 GMT`. In principle this is all it should check when it needs to revalidate.

```
$ curl -I http://build.koreader.rocks/download/dict/gcide.tar.gz
HTTP/1.1 200 OK
Date: Sat, 30 Sep 2023 07:44:14 GMT
Content-Type: text/plain
Content-Length: 14958781
Connection: keep-alive
Last-Modified: Wed, 09 Jan 2019 15:35:31 GMT
ETag: "5c3614c3-e440bd"
Cache-Control: max-age=14400
CF-Cache-Status: MISS
[…]
```

```
$ curl -I http://build.koreader.rocks/download/dict/gcide.tar.gz
HTTP/1.1 200 OK
Date: Sat, 30 Sep 2023 07:49:41 GMT
Content-Type: text/plain
Content-Length: 14958781
Connection: keep-alive
Last-Modified: Wed, 09 Jan 2019 15:35:31 GMT
ETag: "5c3614c3-e440bd"
Cache-Control: max-age=14400
CF-Cache-Status: REVALIDATED
[…]
```

```
$ curl -I http://build.koreader.rocks/download/dict/gcide.tar.gz
HTTP/1.1 200 OK
Date: Sat, 30 Sep 2023 07:50:01 GMT
Content-Type: text/plain
Content-Length: 14958781
Connection: keep-alive
Last-Modified: Wed, 09 Jan 2019 15:35:31 GMT
ETag: "5c3614c3-e440bd"
Cache-Control: max-age=14400
CF-Cache-Status: HIT
[…]
```

```
$ curl -v http://build.koreader.rocks/download/dict/gcide.tar.gz
* processing: http://build.koreader.rocks/download/dict/gcide.tar.gz
*   Trying 188.114.97.3:80...
* Connected to build.koreader.rocks (188.114.97.3) port 80
> GET /download/dict/gcide.tar.gz HTTP/1.1
> Host: build.koreader.rocks
> User-Agent: curl/8.2.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Sat, 30 Sep 2023 07:53:48 GMT
< Content-Type: text/plain
< Content-Length: 14958781
< Connection: keep-alive
< Last-Modified: Wed, 09 Jan 2019 15:35:31 GMT
< ETag: "5c3614c3-e440bd"
< Cache-Control: max-age=14400
< CF-Cache-Status: HIT
```

> I haven't seen cache related headers in the nginx configuration,

Indeed, it might help a few percentage points to explicitly add some for at least a week (except for some files that should only be cached a few hours or maybe a day tops).

But ultimately the bandwidth looks like this. Those are quite comfortable numbers. If I find the time later I'll set up my VPS that I'm not really using because Azure is just… weirdly unattractive, but first it's probably better to identify where the data leak is coming from. But in any case it can't be from build.koreader.rocks or ota.koreader.rocks since as stated those pass completely through CF and are tracked by CF. :-/

![image](https://github.com/koreader/koreader/assets/202757/72eca214-096a-4622-a021-30404c825202)

**Comment by Hzj-jie** [comment]
Ok, my previous company used some edge networking services a while ago which transfered the data across their edge servers, so the static requests would never reach the server again after some 5 minutes I recall.

But anyway, glad to know CF does not need any heads to function. (I remember we discussed this before :)
Though it's still suspicious that build.koreader.rocks is contributing most of the traffic.

**Comment by Frenzie** [comment]
The long and short of it is that it's all proxied and I can't imagine them somehow missing tens of gigabytes of data traffic. 47 (total) - 26 (cached) = 21 GB traffic from build. and ota over the past week.

![image](https://github.com/koreader/koreader/assets/202757/62f2b679-1a39-4452-9451-01d1945f0586)


![image](https://github.com/koreader/koreader/assets/202757/9b7de91f-d7a2-4bf4-a431-aaf1210046bf)

Which it seems to me really only leaves the thing I blanked out here, or something internal to Azure.

Although technically this is possible:
```
curl -I build.koreader.rocks/download --resolve build.koreader.rocks:[[IP]]
```

With some file:
```
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 02 Oct 2023 07:32:54 GMT
Content-Type: application/zip
Content-Length: 42455211
Last-Modified: Sun, 10 Sep 2023 06:10:46 GMT
Connection: keep-alive
ETag: "64fd5de6-287d0ab"
Accept-Ranges: bytes
```

Of course that's not quite as it should be so we'll have to double check the traffic numbers from the nginx logs just in case.

https://developers.cloudflare.com/fundamentals/setup/allow-cloudflare-ip-addresses/

**Comment by Frenzie** [comment]
I've added these caching headers:

```
      location ~* \.(apk|AppImage|targz|tar.gz|zip)$ {
        add_header Cache-Control "public, max-age=86400, immutable";
      }

      location ~* \.zsync$ {
        add_header Cache-Control "public, max-age=3600";
      }
```

Of course as stated this will only potentially affect the 22 GB of uncached traffic (mainly depending on whether CF merely checks last-modified or simply redownloads the entire thing because why not), not whatever traffic is actually being problematic.

![image](https://github.com/koreader/koreader/assets/202757/d2afe7ec-8c29-448c-bf48-4b097840cc38)

**Comment by Frenzie** [comment]
build.koreader.rocks says 10.12 GB over the past 7 days (as in goaccess with `--keep-last=7`)

<img src=https://github.com/koreader/koreader/assets/202757/7d23f801-26d0-4c2a-ad75-9fd6c4a2b0a3 width=40%>

The ota.koreader.rocks logs say **107.58 GB**. That's most peculiar to say the least because I only see CF IPs in the logs. Almost all of it from `172.71.126.something`, and some fairly negligible amounts from other CF IPs. Then *why doesn't CF show this*? It was clearly served through CF. This doesn't make a lick of sense.

<img src=https://github.com/koreader/koreader/assets/202757/f5043757-2eac-471a-8bbb-ce48b88204c8 width=40%>
<img src=https://github.com/koreader/koreader/assets/202757/50b09b54-751e-4886-89cd-aaafcd011f70 width=40%>

**Comment by Hzj-jie** [comment]
That's really a lot. Can cf handle 206 correctly?

By the way, the name asdfsdfa is cool :+1:

**Comment by Frenzie** [comment]
I think I see what you're saying.

1. A partial content request comes in
2. CF grabs the entire file (which may not necessarily be completely idiotic, provided it then caches it for future requests)
3. CF only serves that which is actually requested (hopefully)

2, 3 ⇒ CF has downloaded some 50–70 GB more from the server than it actually served to end users, explaining the missing data.

This is probably fairly testable, though I don't have the time to right this moment.

**Comment by ilyats** [comment]
To whom it may concern, the OTA server is down again. Latest APK on a mirror site is from 01/19/24

**Comment by Hzj-jie** [comment]
It was resumed yesterday.

**Comment by moshin34** [comment]
My sync randomly stopped working a few days ago across all devices.  Is this a global thing?

**Comment by Miladiir** [comment]
ota.koreader.rocks is down starting from atleast yesterday ~08:00 UTC. I can contribute financially or with my time. I have azure and linux server experience if you guys need help.

**Comment by pazos** [comment]
> ota.koreader.rocks is down starting from atleast yesterday ~08:00 UTC. I can contribute financially or with my time. I have azure and linux server experience if you guys need help.

Thanks for the offer :)

The server will be back online the 26. AFAIK it gets killed when the network bandwitdh reaches some limit. I think @Frenzie  and @Hzj-jie are the people involved with the server management. Not sure what can be done without increasing quotas (I'm a happily unaware :p)

**Comment by Frenzie** [comment]
> Not sure what can be done without increasing quotas (I'm happily unaware :p)

I find the VPS very bad compared to all others I have access to. It's slow and SSH connections are unreliable. From e.g. AWS EC2 I know that's definitely not an issue with being in North America. These issues might be a worthy trade-off for a lot of bandwidth, yet bandwidth is apparently also more than an order of magnitude less than anything else. I am nonetheless grateful for its existence.

But as you know these things largely work on annoyance thresholds and it never quite managed to make enough impact, although it's come close. I definitely wouldn't want to lower the annoyance threshold by rewarding Azure for their terrible product.

**Comment by kyxap** [comment]
btw github pipeline does not much build requirements? just curious

**Comment by pazos** [comment]
> btw github pipeline does not much build requirements? just curious

No idea, but lets focus on solve the bandwidth issue here.

Feel free to open another ticket if you want to build artifacts from github. I think gitlab is better. Not all the eggs on the same basket :)

**Comment by SilverGreen93** [comment]
ota.koreader.rocks is down since a couple of days.

**Comment by Frenzie** [comment]
Closing this one, but see #15902.
